pkg/serviceaccounts/controllers/create_dockercfg_secrets.go

```diff
@@ -0,0 +1,343 @@
+		// we do need to clean up our dockercfgSecret
+		if err := e.client.Secrets(dockercfgSecret.Namespace).Delete(dockercfgSecret.Name); err != nil {
+			glog.Error(err)
+		}
+		return nil
+	}
+
+	return err
+}
+
+// createDockerPullSecretReference updates a service account to reference the dockercfgSecret as a Secret and an ImagePullSecret
+func (e *DockercfgController) createDockerPullSecretReference(staleServiceAccount *api.ServiceAccount, dockercfgSecretName string) ( /*isStale*/ bool, error) {
+	liveServiceAccount, err := e.client.ServiceAccounts(staleServiceAccount.Namespace).Get(staleServiceAccount.Name)
+	if err != nil {
+		return false, err
+	}
+
+	mountableSecrets, imagePullSecrets := getSecretNames(liveServiceAccount)
+	staleMountableSecrets, staleImagePullSecrets := getSecretNames(staleServiceAccount)
+
+	// if we're trying to create a reference based on stale data, let the caller know
+	if !reflect.DeepEqual(staleMountableSecrets.List(), mountableSecrets.List()) || !reflect.DeepEqual(staleImagePullSecrets.List(), imagePullSecrets.List()) {
+		return true, fmt.Errorf("cannot add reference based on stale data.  decision made for %v, but live version is %v", staleServiceAccount.ResourceVersion, liveServiceAccount.ResourceVersion)
+	}
+
+	changed := false
+	if !mountableSecrets.Has(dockercfgSecretName) {
+		liveServiceAccount.Secrets = append(liveServiceAccount.Secrets, api.ObjectReference{Name: dockercfgSecretName})
+		changed = true
+	}
+
+	if !imagePullSecrets.Has(dockercfgSecretName) {
+		liveServiceAccount.ImagePullSecrets = append(liveServiceAccount.ImagePullSecrets, api.LocalObjectReference{Name: dockercfgSecretName})
+		changed = true
+	}
+
+	if changed {
+		if _, err = e.client.ServiceAccounts(liveServiceAccount.Namespace).Update(liveServiceAccount); err != nil {
+			return false, err
+		}
+	}
+	return false, nil
+}
+
+const (
+	tokenSecretWaitInterval = 100 * time.Millisecond
+	tokenSecretWaitTimes    = 20
+)
+
+// createTokenSecret creates a token secret for a given service account.  Returns the name of the token
+func (e *DockercfgController) createTokenSecret(serviceAccount *api.ServiceAccount) (*api.Secret, error) {
+	tokenSecret := &api.Secret{
+		ObjectMeta: api.ObjectMeta{
+			Name:      secret.Strategy.GenerateName(getTokenSecretNamePrefix(serviceAccount)),
+			Namespace: serviceAccount.Namespace,
+			Annotations: map[string]string{
+				api.ServiceAccountNameKey: serviceAccount.Name,
+				api.ServiceAccountUIDKey:  string(serviceAccount.UID),
+			},
+		},
+		Type: api.SecretTypeServiceAccountToken,
+		Data: map[string][]byte{},
+	}
+
+	_, err := e.client.Secrets(tokenSecret.Namespace).Create(tokenSecret)
+	if err != nil {
+		return nil, err
+	}
+
+	// now we have to wait for the service account token controller to make this valid
+	// TODO remove this once we have a create-token endpoint
+	for i := 0; i <= tokenSecretWaitTimes; i++ {
```

Comment:
yeah, this uses a deltafifo, so add/update notifications get serialized... makes the token creation far more important